etc/mview.xml

```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:framework:Mview/etc/mview.xsd">
+    <view id="async_event" class="Aligent\AsyncEvents\Model\Indexer\Mview\Action" group="indexer">
```

Comment:
Rather than having a separate class that instantiates your indexer and calls it, you can just have your indexer implement the interface.

Reply:
Actually, I see it already does implement, so you can just point to it here

Reply:
Great pickup. Actually that was the way it was implemented. The only example that indexes documents into Elasticsearch is `\Magento\CatalogSearch\Model\Indexer\Fulltext`
which has some optional DI parameters in its constructor. 

When the `bin/magento cron:run` boots up `indexer_update_all_views`, the core one implements a separate class, which in turn uses a factory to instantiate the indexer.

```php
/** @var \Magento\Framework\Indexer\IndexerInterface $indexer */
$indexer = $this->indexerFactory->create()->load(Fulltext::INDEXER_ID);
$indexer->reindexList($ids);
```

This works because `reindexList` will instantiate the indexer using

```php
/**
 * Return indexer action instance
 *
 * @return ActionInterface
 * @throws \InvalidArgumentException
 */
protected function getActionInstance()
{
    return $this->actionFactory->create(
        $this->getActionClass(),
        [
            'indexStructure' => $this->getStructureInstance(),
            'data' => $this->getData(),
        ]
    );
}
```

Pointing the indexer class directly in the `mview.xml` meant that when it's trying to instantiate the class it runs through


```php
// \Magento\Framework\Mview\View::update

$action = $this->actionFactory->get($this->getActionClass());
```

This kept trying to instantiate the below (note the array)

```php
    public function __construct(
        DimensionProviderInterface $dimensionProvider,
        IndexerHandlerFactory $indexerHandlerFactory,
        AsyncEventSubscriberLogs $asyncEventSubscriberLogsDataProvider,
        AsyncEvent $asyncEventScopeResolver,
        array $data,
        int $batchSize = null,
        DeploymentConfig $deploymentConfig = null
    )
```

and would simply just fail with

```php
[2022-08-14 10:13:41] report.CRITICAL: Missing required argument $data of Aligent\AsyncEvents\Model\Indexer\AsyncEventSubscriber. {"exception":"[object] (BadMethodCallException(code: 0): Missing required argument $data of Aligent\\AsyncEvents\\Model\\Indexer\\AsyncEventSubscriber. at /app/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:264)"} []
```

I couldn't find a nice way to simply pass `$data` myself. It is crucial for instantiating the Elasticsearch indexer handler.

```php
// \Aligent\AsyncEvents\Model\Indexer\AsyncEventSubscriber::executeByDimensions
$saveHandler = $this->indexerHandlerFactory->create(
    [
        'data' => $this->data,
        'scopeResolver' => $this->asyncEventScopeResolver
    ]
);

```

If there's a better way instead of the proxied factory approach I'd be happy to implement it.

Reply:
@aligent-lturner are you ok to mark this as resolved?